// @appinsights/aichartcore/AI/ME/Utilities/CallbackRegistrationMap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var CallbackRegistrationMap = /** @class */ (function () {
    function CallbackRegistrationMap() {
        this._map = {};
    }
    CallbackRegistrationMap.prototype.Register = function (registrationInfo) {
        if (this._map[registrationInfo.id]) {
            throw new Error("Error: Reregistering an existing id is not allowed. id = " + registrationInfo.id);
        }
        else {
            this._map[registrationInfo.id] = registrationInfo;
        }
    };
    CallbackRegistrationMap.prototype.Unregister = function (id) {
        delete this._map[id];
    };
    CallbackRegistrationMap.prototype.ExecuteAction = function (id, action) {
        for (var key in this._map) {
            if (this._map.hasOwnProperty(key) && key !== id) {
                var element = this._map[key];
                action(element);
            }
        }
    };
    /**
     * Indicating whether the map is empty meaning that there's no subscriber register with it.
     */
    CallbackRegistrationMap.prototype.IsEmpty = function () {
        return Object.keys(this._map).length === 0;
    };
    /**
     * For testing purpose
     */
    CallbackRegistrationMap.prototype.GetCallbackRegistrationInfo = function (id) {
        return this._map[id];
    };
    return CallbackRegistrationMap;
}());
exports.CallbackRegistrationMap = CallbackRegistrationMap;
