// appinsights-iframe-shared/out/Contracts/CuratedContracts.js
export var KQLOperator;
(function (KQLOperator) {
    KQLOperator["None"] = "";
    KQLOperator["Equals"] = "==";
    KQLOperator["NotEqualTo"] = "!=";
    KQLOperator["GreaterThan"] = ">";
    KQLOperator["GreaterThanOrEqualTo"] = ">=";
    KQLOperator["LessThan"] = "<";
    KQLOperator["LessThanOrEqualTo"] = "<=";
    KQLOperator["In"] = "in";
    KQLOperator["NotIn"] = "!in";
})(KQLOperator || (KQLOperator = {}));
export var Experience;
(function (Experience) {
    Experience[Experience["Requests"] = 1] = "Requests";
    Experience[Experience["Dependencies"] = 2] = "Dependencies";
    Experience[Experience["Exceptions"] = 3] = "Exceptions";
    Experience[Experience["Servers"] = 4] = "Servers";
    Experience[Experience["AvailabilityResults"] = 5] = "AvailabilityResults";
})(Experience || (Experience = {}));
export var TrendPercentile;
(function (TrendPercentile) {
    TrendPercentile[TrendPercentile["Average"] = 1] = "Average";
    TrendPercentile[TrendPercentile["Percentile50"] = 2] = "Percentile50";
    TrendPercentile[TrendPercentile["Percentile95"] = 3] = "Percentile95";
    TrendPercentile[TrendPercentile["Percentile99"] = 4] = "Percentile99";
})(TrendPercentile || (TrendPercentile = {}));
export var ClientTypeMode;
(function (ClientTypeMode) {
    ClientTypeMode["Server"] = "Server";
    ClientTypeMode["Browser"] = "Browser";
})(ClientTypeMode || (ClientTypeMode = {}));
export var FilterableItemKind;
(function (FilterableItemKind) {
    FilterableItemKind[FilterableItemKind["Operation"] = 1] = "Operation";
    FilterableItemKind[FilterableItemKind["Exception"] = 2] = "Exception";
    FilterableItemKind[FilterableItemKind["Dependency"] = 3] = "Dependency";
    FilterableItemKind[FilterableItemKind["AvailabilityResult"] = 4] = "AvailabilityResult";
})(FilterableItemKind || (FilterableItemKind = {}));
