// appinsights-iframe-shared/out/Components/insights/auto-cluster/AutoClusterResultFactory.js
import { AutoClusterResult } from './AutoClusterResult';
var Columns = /** @class */ (function () {
    function Columns() {
    }
    Columns.SegmentId = 'SegmentId';
    Columns.Count = 'Count';
    Columns.Percent = 'Percent';
    return Columns;
}());
export { Columns };
/** Creates auto-cluster results objects. */
var AutoClusterResultFactory = /** @class */ (function () {
    function AutoClusterResultFactory() {
    }
    AutoClusterResultFactory.create = function (rawResult) {
        var count;
        var percent;
        var segmentId;
        var dimensions = {};
        Object.keys(rawResult).forEach(function (column) {
            switch (column) {
                case Columns.Count:
                    count = rawResult[column];
                    break;
                case Columns.Percent:
                    percent = rawResult[column];
                    break;
                case Columns.SegmentId:
                    segmentId = rawResult[column];
                    break;
                default:
                    if (rawResult[column]) {
                        dimensions[column] = rawResult[column];
                    }
                    break;
            }
        });
        if (Object.keys(dimensions).length === 0) {
            return null;
        }
        return new AutoClusterResult({
            id: segmentId,
            count: count,
            percent: percent,
            dimensions: dimensions
        });
    };
    return AutoClusterResultFactory;
}());
export { AutoClusterResultFactory };
