// appinsights-iframe-shared/out/Components/insights/auto-cluster/AutoClusterResultTestUtils.spec.js
import { AutoClusterResult } from './AutoClusterResult';
var AutoClusterResultTestUtils = /** @class */ (function () {
    function AutoClusterResultTestUtils() {
    }
    AutoClusterResultTestUtils.getRawAutoClusterResult = function (segmentId) {
        if (segmentId === void 0) { segmentId = '1'; }
        return {
            Count: 1568,
            Percent: 57.90434,
            SegmentId: segmentId,
            resultCode: '200',
            client_CountryOrRegion: 'Brazil',
            client_City: '',
            client_StateOrProvince: undefined,
            client_OS: null
        };
    };
    AutoClusterResultTestUtils.getAutoClusterResult = function (id) {
        if (id === void 0) { id = '1'; }
        var dimensions = {};
        dimensions['resultCode'] = '200';
        dimensions['client_CountryOrRegion'] = 'Brazil';
        return new AutoClusterResult({
            id: id,
            percent: 57.90434,
            count: 1568,
            dimensions: dimensions
        });
    };
    return AutoClusterResultTestUtils;
}());
export { AutoClusterResultTestUtils };
