// @appinsights/aichartcore/AI/ME/Skins/ThresholdSkinsShared.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ThresholdSkinsShared;
(function (ThresholdSkinsShared) {
    var DynamicThresholdOperator;
    (function (DynamicThresholdOperator) {
        DynamicThresholdOperator[DynamicThresholdOperator["Both"] = 1] = "Both";
        DynamicThresholdOperator[DynamicThresholdOperator["Up"] = 2] = "Up";
        DynamicThresholdOperator[DynamicThresholdOperator["Down"] = 3] = "Down";
    })(DynamicThresholdOperator = ThresholdSkinsShared.DynamicThresholdOperator || (ThresholdSkinsShared.DynamicThresholdOperator = {}));
    var DynamicThresholdSensitivity;
    (function (DynamicThresholdSensitivity) {
        DynamicThresholdSensitivity[DynamicThresholdSensitivity["Low"] = 1] = "Low";
        DynamicThresholdSensitivity[DynamicThresholdSensitivity["Medium"] = 2] = "Medium";
        DynamicThresholdSensitivity[DynamicThresholdSensitivity["High"] = 3] = "High";
    })(DynamicThresholdSensitivity = ThresholdSkinsShared.DynamicThresholdSensitivity || (ThresholdSkinsShared.DynamicThresholdSensitivity = {}));
    var DynamicThresholdFlow;
    (function (DynamicThresholdFlow) {
        DynamicThresholdFlow[DynamicThresholdFlow["GetPredictions"] = 1] = "GetPredictions";
        DynamicThresholdFlow[DynamicThresholdFlow["CalculateBaseline"] = 2] = "CalculateBaseline";
    })(DynamicThresholdFlow = ThresholdSkinsShared.DynamicThresholdFlow || (ThresholdSkinsShared.DynamicThresholdFlow = {}));
    ThresholdSkinsShared.DynamicThresholdDefaultValues = {
        calculateBaselineStartTimeOffsetInDays: 10,
        consecutiveViolations: 2,
        consecutivePeriods: 2,
        flow: DynamicThresholdFlow.CalculateBaseline,
        sensitivity: DynamicThresholdSensitivity.Medium,
        startTimeEnabled: false,
        operator: DynamicThresholdOperator.Both,
        isEligible: false,
        responseMessage: "",
        isDynamic: false,
        modelId: "",
    };
})(ThresholdSkinsShared = exports.ThresholdSkinsShared || (exports.ThresholdSkinsShared = {}));
