// appinsights-iframe-shared/out/Components/dropdown/DropdownTypes.js
var DropdownDivider = /** @class */ (function () {
    function DropdownDivider(id, displayName, icon) {
        this.id = id;
        this.displayName = displayName;
        this.icon = icon;
    }
    return DropdownDivider;
}());
export { DropdownDivider };
var DropdownLink = /** @class */ (function () {
    function DropdownLink(id, displayName, url, icon) {
        this.id = id;
        this.displayName = displayName;
        this.url = url;
        this.icon = icon;
    }
    return DropdownLink;
}());
export { DropdownLink };
var DropdownCheckbox = /** @class */ (function () {
    function DropdownCheckbox(id, displayName, url, icon) {
        this.id = id;
        this.displayName = displayName;
        this.url = url;
        this.icon = icon;
    }
    return DropdownCheckbox;
}());
export { DropdownCheckbox };
var DropdownMessage = /** @class */ (function () {
    function DropdownMessage(id, displayName, message, data, icon) {
        this.id = id;
        this.displayName = displayName;
        this.message = message;
        this.data = data;
        this.icon = icon;
    }
    return DropdownMessage;
}());
export { DropdownMessage };
var DropdownAction = /** @class */ (function () {
    function DropdownAction(id, displayName, dispatchActionCallback, icon, role, dontClose, checkboxState) {
        this.id = id;
        this.displayName = displayName;
        this.dispatchActionCallback = dispatchActionCallback;
        this.icon = icon;
        this.checkboxState = checkboxState;
        this.role = role;
        this.dontCloseOnSelect = dontClose;
    }
    return DropdownAction;
}());
export { DropdownAction };
