// @appinsights/pillscontrol-es5/dist/TimeInterfaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var TimeValues;
(function (TimeValues) {
    TimeValues[TimeValues["Last30Minutes"] = 1800000] = "Last30Minutes";
    TimeValues[TimeValues["LastHour"] = 3600000] = "LastHour";
    TimeValues[TimeValues["Last4Hours"] = 14400000] = "Last4Hours";
    TimeValues[TimeValues["Last6Hours"] = 21600000] = "Last6Hours";
    TimeValues[TimeValues["Last12Hours"] = 43200000] = "Last12Hours";
    TimeValues[TimeValues["LastDay"] = 86400000] = "LastDay";
    TimeValues[TimeValues["Last2Days"] = 172800000] = "Last2Days";
    TimeValues[TimeValues["Last3Days"] = 259200000] = "Last3Days";
    TimeValues[TimeValues["Last7Days"] = 604800000] = "Last7Days";
    TimeValues[TimeValues["Last30Days"] = 2592000000] = "Last30Days";
    TimeValues[TimeValues["Last90Days"] = 7776000000] = "Last90Days";
    TimeValues[TimeValues["Custom"] = 0] = "Custom";
})(TimeValues = exports.TimeValues || (exports.TimeValues = {}));
/**
 * Time grains stored according to the ISO 8601 duration spec
 * https://en.wikipedia.org/wiki/ISO_8601#Durations
 */
var TimeGrains;
(function (TimeGrains) {
    TimeGrains[TimeGrains["Auto"] = 0] = "Auto";
    TimeGrains[TimeGrains["Minute"] = 1] = "Minute";
    TimeGrains[TimeGrains["FiveMinute"] = 2] = "FiveMinute";
    TimeGrains[TimeGrains["TenMinute"] = 3] = "TenMinute";
    TimeGrains[TimeGrains["FifteenMinute"] = 4] = "FifteenMinute";
    TimeGrains[TimeGrains["ThirtyMinute"] = 5] = "ThirtyMinute";
    TimeGrains[TimeGrains["Hour"] = 6] = "Hour";
    TimeGrains[TimeGrains["SixHours"] = 7] = "SixHours";
    TimeGrains[TimeGrains["TwelveHours"] = 8] = "TwelveHours";
    TimeGrains[TimeGrains["Day"] = 9] = "Day";
    TimeGrains[TimeGrains["Week"] = 10] = "Week";
    TimeGrains[TimeGrains["Month"] = 11] = "Month";
})(TimeGrains = exports.TimeGrains || (exports.TimeGrains = {}));
