// @appinsights/aichartcore/Scripts/Strings/src/MELiteStrings.js
"use strict";
/**
 * To generate Strings.js and Strings.d.ts files -
 * run "tsc MELiteStrings.ts -m "CommonJS" -d" in this folder
 * then copy MELiteStrings.js file and MELiteStrings.d.ts to
 * Strings.js and Strings.d.ts(overwrite) in the folder outside src;
 * then run "yarn install --force" to install the new npm package;
 * remember to checkin all 3 files MELiteStrings.ts, Strings.d.ts and Strings.js
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * TODO: The contents of this file need to be localized once our system supports it.
 */
exports.categoryClassicLabel = " (classic)";
/**
 * Category names for metrics we group on the client.
 */
exports.DraftCategoryDisplayName = {
    availability: "Availability",
    usage: "Usage",
    server: "Server",
    failure: "Failure",
    performanceCounter: "Performance Counter",
    client: "Client",
    customEvent: "Custom",
};
/**
 * When a segment name is not present or a metric display name is not specified, we default to this.
 */
exports.undefinedMetricOrSegment = "<UNDEFINED>";
/**
 * Messages to display to the user when there is NoData,ErrorRetrievingData, Throttled, etc.
 */
exports.errorRetrievingDataMessage = "Error retrieving data.";
exports.throttledMessage = "Please refresh."; // 429
exports.unauthorizedMessage = "Session expired."; // 401
exports.forbiddenMessage = "Access permission denied."; // 403
exports.resourceNotfoundMessage = "Resource not found."; // 404
/**
 * Text that is presented to the user when they have no data in their chart because backend returned nothing.
 */
exports.noDataForMetric = "No data in the given time range";
exports.noRolesMessage = "No roles configured.";
/**
 * Text that is presented to the user when they data points for a chart series exceed the MaxiumDataPointsPerSeries.
 */
exports.tooManyDataPoints = "The chart contains more than 10,000 data points. Please choose a smaller time range or coarse time grain.";
exports.noCommonGrain = "Failed to find a common supported grain for given metrics.";
exports.grainNotSupported = "Selected grain is not supported by metric(s).";
exports.dimensionNotSupported = "Selected dimension(s) are not supported by metric(s).";
exports.supportedDimensions = "Supported ones are {0}.";
/**
 * Display name for aggregation drop down, make sure the index here match the value in Constants.Aggregation enum,
 */
exports.AggregationDisplayString = [
    "None",
    "Sum",
    "Min",
    "Max",
    "Avg",
    "Unique",
    "Percentile",
    "Count",
    "Last",
];
exports.TimeUnitsDisplay = {
    ticks: "ticks",
    milliseconds: "ms",
    seconds: "sec",
    minutes: "min",
    hours: "hours",
    days: "days",
    weeks: "weeks",
    months: "months",
    years: "years",
};
exports.billionUnit = "B";
exports.millionUnit = "M";
exports.kiloUnit = "k";
/**
 * Draft document type display names
 */
exports.documentTypeRequest = "Request";
exports.documentTypeTrace = "Trace";
exports.documentTypePageView = "Page View";
exports.documentTypeCustom = "Custom Event";
exports.documentTypeException = "Exception";
exports.documentTypeDependency = "Dependency";
exports.documentTypeAvailability = "Availability";
exports.documentTypeBrowserTimings = "Browser Timing";
exports.documentTypePerf = "Performance Counter";
exports.documentTypeBillingMeters = "Billing Meters";
exports.aILiveMetric = "Live Metric";
exports.aIHistorical = "Historical";
/**
 * Chart title related text...
 */
exports.chartTitleWithGrouping = "{0} by {1}";
exports.chartTitleWithGroupingFilters = "{0} by {1} where {2}";
exports.chartTitleWithFilters = "{0} where {1}";
exports.chartTitleTwoItems = "{0} and {1}";
exports.chartTitleThreeItems = "{0}, {1}, and {2}";
exports.chartTitleThreePlusItems = "{0}, {1}, and {2} other {3}";
exports.chartTitleMetrics = "metrics";
exports.chartTitleResources = "resources";
exports.chartTitleFor = " for ";
exports.perSecond = "/s";
exports.legendNoValueForNarrator = "No value at {0}"; // No value at Mon Apr 23 2018 07:35:00 GMT-0700 (Pacific Daylight Time)
exports.dataPointForNarrator = "{0} {1} at {2}"; // "2.4 GiB at Mon Apr 23 2018 07:35:00 GMT-0700 (Pacific Daylight Time)
exports.chartKeyboardShortcutDescription = "To Interact with the chart with the keyboard, use left/right arrows to move between series, and up/down arrows to move between data points.";
// QPS Schema friendly names -
exports.requestCountFriendlyName = "Request Rate";
exports.requestDurationFriendlyName = "Request Duration";
exports.failedRequestsFriendlyName = "Request Failure Rate";
exports.dependencyCountFriendlyName = "Dependency Call Rate";
exports.dependencyDurationFriendlyName = "Dependency Call Duration";
exports.failedDependenciesFriendlyName = "Dependency Call Failure Rate";
exports.availableMemoryFriendlyName = "Committed Memory";
exports.processCPUFriendlyName = "CPU Total (%)";
exports.exceptionCountFriendlyName = "ExceptionRate";
exports.contributorsCountDefinitionFriendlyName = "Contributors Count";
exports.webAppCpuTitleString = "Process CPU (sum of % across all cores)";
