// appinsights-iframe-shared/out/Components/toggle/Toggle.js
import * as React from 'react';
import { RadioGroup } from '..';
export var Toggle = function (_a) {
    var selected = _a.selected, leftOption = _a.leftOption, rightOption = _a.rightOption, onSelected = _a.onSelected, label = _a.label;
    function getToggleOption(option, className) {
        if (option.icon) {
            return React.createElement("div", { className: className + (option.value === selected ? ' selected' : ''), title: option.text },
                React.createElement("div", { className: 'toggle-screen' }),
                option.icon);
        }
        else if (option.text) {
            return React.createElement("div", { className: className + (option.value === selected ? ' selected' : '') },
                React.createElement("div", { className: 'toggle-screen' }),
                option.text);
        }
        else {
            return React.createElement("div", null);
        }
    }
    var leftId = leftOption.text;
    var rightId = rightOption.text;
    if (leftId && rightId) {
        return React.createElement(RadioGroup, { className: 'toggle noselect', items: [
                {
                    content: getToggleOption(leftOption, 'left-option'),
                    id: leftId,
                    title: leftOption.text,
                    contentId: leftId,
                    value: leftOption,
                    className: leftOption.icon ? ' with-icon' : ''
                },
                {
                    content: getToggleOption(rightOption, 'right-option'),
                    id: rightId,
                    title: rightOption.text,
                    contentId: rightId,
                    value: rightOption,
                    className: rightOption.icon ? ' with-icon' : ''
                }
            ], label: label, selectedIndex: leftOption.value === selected ? 0 : 1, onChange: function (value) { return onSelected(value.value); } });
    }
    else {
        return React.createElement("div", null);
    }
};
