// appinsights-iframe-shared/out/Components/FeedbackDropdown.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { DropdownWithLinks, DropdownLink, DropdownMessage, ComboBoxHeader } from './dropdown';
import { SmileHappySvg, SmileSadSvg, ExternalLinkSvg, Utils } from '../';
var FeedbackDropdown = /** @class */ (function (_super) {
    tslib_1.__extends(FeedbackDropdown, _super);
    function FeedbackDropdown(props) {
        var _this = _super.call(this, props) || this;
        _this._onSelected = _this.onSelected.bind(_this);
        return _this;
    }
    FeedbackDropdown.prototype.render = function () {
        var displayName = Utils.GetValueOrDefault(this.props.strings, 'feedbackDropdown', '');
        return React.createElement(DropdownWithLinks, { flyoutClassName: 'dropdown-flyout right-flyout-direction', role: this.props.role, tabIndex: this.props.tabIndex, isFocused: this.props.isFocused, notPartOfMenu: this.props.notPartOfMenu, wrapperClassName: 'combobox-dropdown-wrapper ', onChange: this._onSelected, label: displayName, header: React.createElement(ComboBoxHeader, { displayName: displayName, icon: React.createElement(SmileHappySvg, null) }), messageService: this.props.messageService, options: [
                new DropdownMessage('SendSmile', this.props.strings['sendSmile'], this.props.message, { rating: '5' }, React.createElement(SmileHappySvg, null)),
                new DropdownMessage('SendFrown', this.props.strings['sendFrown'], this.props.message, { rating: '1' }, React.createElement(SmileSadSvg, null)),
                new DropdownLink('SendFeedback', this.props.strings['sendFeedback'], 'https://feedback.azure.com/forums/357324-application-insights/', React.createElement(ExternalLinkSvg, null)),
            ] });
    };
    FeedbackDropdown.prototype.onSelected = function (selected) {
        this.props.telemetryContext.RecordClick(selected.id, null);
    };
    FeedbackDropdown.defaultProps = {
        isFocused: false,
        role: 'menu',
        tabIndex: 0
    };
    return FeedbackDropdown;
}(React.Component));
export { FeedbackDropdown };
