// appinsights-iframe-shared/out/Components/zoom-button/ZoomButton.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { AccessibilityUtils } from '../../AccessibilityUtils';
var ZoomButton = /** @class */ (function (_super) {
    tslib_1.__extends(ZoomButton, _super);
    function ZoomButton(props) {
        var _this = _super.call(this, props) || this;
        _this._onKeyDown = _this.onKeyDown.bind(_this);
        _this._onKeyUp = _this.onKeyUp.bind(_this);
        _this.state = {
            mouseDown: false
        };
        return _this;
    }
    ZoomButton.prototype.render = function () {
        return React.createElement("div", tslib_1.__assign({ className: 'zoom-button' + (this.state.mouseDown ? ' mouse-down' : ''), role: 'button' }, AccessibilityUtils.getAccessibilityProps({
            onClick: this.props.onClick,
        }), { onKeyDown: this._onKeyDown, onKeyUp: this._onKeyUp, title: this.props.title }),
            React.createElement("svg", { className: 'outline' },
                React.createElement("circle", { r: 18, cx: 19, cy: 19 }),
                React.createElement("circle", { className: 'highlight', r: 18, cx: 19, cy: 19 }),
                React.createElement("svg", { height: 16, width: 16, x: 12, y: 12, className: 'icon' }, this.props.icon)));
    };
    ZoomButton.prototype.onKeyDown = function (e) {
        if (e.key === 'Enter' || e.key === ' ') {
            this.setState({ mouseDown: true });
        }
    };
    ZoomButton.prototype.onKeyUp = function (e) {
        if (e.key === 'Enter' || e.key === ' ') {
            this.setState({ mouseDown: false });
        }
    };
    return ZoomButton;
}(React.PureComponent));
export { ZoomButton };
