// appinsights-iframe-shared/out/Components/insights/InsightsTestUtils.spec.js
import * as tslib_1 from "tslib";
import * as React from 'react';
import { CardResourcesProviderBase } from './CardResourcesProviderBase';
var DummyCardResourcesProvider = /** @class */ (function (_super) {
    tslib_1.__extends(DummyCardResourcesProvider, _super);
    function DummyCardResourcesProvider(strings, detection) {
        return _super.call(this, strings, detection) || this;
    }
    DummyCardResourcesProvider.prototype.getCardHeader = function () {
        var props = {
            icon: null,
            title: React.createElement('div', null, 'Dummy title'),
            subtitle: 'Dummy sub title',
            isLoading: false,
            ariaLabel: 'Dummy title'
        };
        return props;
    };
    DummyCardResourcesProvider.prototype.getCardContent = function () {
        return React.createElement('div');
    };
    return DummyCardResourcesProvider;
}(CardResourcesProviderBase));
export { DummyCardResourcesProvider };
var InsightsTestUtils = /** @class */ (function () {
    function InsightsTestUtils() {
    }
    InsightsTestUtils.getStrings = function () {
        return {
            'disttributionTitle': 'Distribution',
            'smartDetectionOptionTitle': 'Insights',
            'timeSeriesCardTitleForDetectionWithSpecificTargetValue': 'Degradation in {0} for "{1}"',
            'timeSeriesCardTitleForDependencyWithSingleNameDetection': 'Degradation in Dependency duration for "{0}" {1} calls to "{2}"',
            'timeSeriesCardTitleForDependencyWithSingleNameDetectionOmitBaseName': 'Degradation in Dependency duration for "{0}" {1} calls',
            'timeSeriesCardTitleForDependencyWitMultipleNamesDetection': 'Degradation in dependency Duration for multiple {0} calls to {1}',
            'timeSeriesCardWhatPropertyTitle': '{0} {1} vs {2} in the previous 7 days',
            'operationNameDimensionDisplayName': 'Operation name',
            'baseNameDimensionDisplayName': 'Base name',
            'dependenciesDuration': 'duration',
            'serverResponseTime': 'Server response time',
            'descriptionTitle': 'What:',
            'eventDateTitle': 'When:',
            'noteLabel': 'Note:',
            'thingsToNoteNoImpactedUserText': '{0}% of all {1} were affected:',
            'thingsToNoteSingleImpactedUserText': '1 user and {0}% of all {1} were affected :',
            'thingsToNoteMultipleImpactedUsersText': '{0} users and {1}% of all {2} were affected:',
            'viewDetails': 'View details',
            'responseTimesByOperationNameDescription': 'Slow response time for "{0}" Server requests',
            'dependencyTimesByDependencyNamePartitionedDescription': 'Long Dependency duration for "{1}" {0} calls',
            'responseTimesByMultiplePredicatesDescription': 'Slow response time for Server requests with these properties:',
            'dependencyTimesByMultiplePredicatesDescription': 'Slow response to dependency calls with these parameters:',
            'requestWhatTemplate': '{0} average response time vs {1} for all other requests',
            'dependencyPartitionedWhatTemplate': '{0} average response time vs {1} for all other {2} calls',
            'dependencyWhatTemplateByDependencytype': '{0} average response time vs {1} for all other dependency types',
            'serverRequests': 'Server requests',
            'dependencyCalls': 'Dependency calls',
            'operatingSystem': 'Operating system',
            'noInsightsMessage': 'No insights were found',
            'timeSeriesCardWhatForDependencyWithMultipleSegments': 'Dependency duration is longer than in previous 7 days',
            'smartDetectionCardAverageTimeLabelText': '{0} over previous 7 days',
            'smartDetectionCardTitle': '{0} degradation:',
            'responseTime': 'response time',
            'duration': 'duration',
            'defaultOperationName': 'Overall',
            'select': 'Select {0}',
            'operation': 'operation',
            'dependencyWithLower': 'dependency',
            'anomalyDetectionCardTitle': 'Longer {0}:',
            'anomalyDetectionAllOtherLabel': 'All other {0}:',
            'operationsTitle': 'operations',
            'smartDetectionCardAverageTimeLabelTitle': 'Daily Avg. time:',
            'smartDetectionCardDetectedTimeLabelTitle': 'Detected time:',
            'xOutOfY': '{0} out of {1}',
            'autoClusterCardHeaderText': 'Common properties:'
        };
    };
    return InsightsTestUtils;
}());
export { InsightsTestUtils };
