// appinsights-iframe-shared/out/Services/Query/DraftDataTypes.js
export var ODataKeys = ['timespan', '$filter', '$search', '$orderby',
    '$select', '$skip', '$top', '$format', '$count', '$apply'];
export var HeaderNames = {
    Authorization: 'Authorization',
    SessionId: 'x-ms-client-session-id',
    App: 'x-ms-app',
    ClientRequestId: 'x-ms-client-request-id',
    UserId: 'x-ms-user-id',
    ClientRequestInfo: 'x-ms-client-request-info',
    CacheControl: 'Cache-Control',
    HostingBladeInstanceId: 'x-ms-client-blade-instance-id',
    RetryAfter: 'Retry-After'
};
export var HeaderValues = {
    // todo: should we have a separate app name?
    App: 'AppInsightsPortalExtension'
};
export var Paths = {
    KQLQuery: 'https://{0}/{2}/apps/{1}/query',
    ODataQuery: 'https://{0}/{2}/apps/{1}/events',
    MetricQuery: 'https://{0}/{2}/apps/{1}/metrics',
    MetricMetadataQuery: 'https://{0}/{2}/apps/{1}/metrics/metadata',
    Metadata: 'https://{0}/{2}/apps/{1}/metadata'
};
export var Versions = {
    Beta: 'beta',
    V1: 'v1'
};
export var StatusCodes = {
    Success: 200,
    NoContent: 204,
    TooManyRequests: 429,
    BadRequest: 400
};
export var DraftQueryBackend;
(function (DraftQueryBackend) {
    DraftQueryBackend[DraftQueryBackend["Kusto"] = 1] = "Kusto";
    DraftQueryBackend[DraftQueryBackend["MDM"] = 2] = "MDM";
})(DraftQueryBackend || (DraftQueryBackend = {}));
